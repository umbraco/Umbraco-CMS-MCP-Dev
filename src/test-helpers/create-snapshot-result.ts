import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
