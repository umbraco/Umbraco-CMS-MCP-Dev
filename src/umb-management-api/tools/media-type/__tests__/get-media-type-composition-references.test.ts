import GetMediaTypeCompositionReferencesTool from "../get/get-media-type-composition-references.js";
import { MediaTypeBuilder } from "./helpers/media-type-builder.js";
import { MediaTypeTestHelper } from "./helpers/media-type-helper.js";
import { jest } from "@jest/globals";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";

const TEST_MEDIA_TYPE_NAME = "_Test MediaType Composition";
const TEST_COMPOSITION_NAME = "_Test Composition MediaType";

describe("get-media-type-composition-references", () => {
  let originalConsoleError: typeof console.error;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(async () => {
    console.error = originalConsoleError;
    // Clean up any test media types
    await MediaTypeTestHelper.cleanup(TEST_MEDIA_TYPE_NAME);
    await MediaTypeTestHelper.cleanup(TEST_COMPOSITION_NAME);
  });

  it("should get composition references for a media type", async () => {
    // Create a media type to be used as a composition
    const compositionBuilder = await new MediaTypeBuilder()
      .withName(TEST_COMPOSITION_NAME)
      .create();

    // Create a parent media type that uses the composition
    const parentBuilder = await new MediaTypeBuilder()
      .withName(TEST_MEDIA_TYPE_NAME)
      .withComposition(compositionBuilder.getId())
      .create();

    // Get the composition references
    const result = await GetMediaTypeCompositionReferencesTool().handler({
      id: compositionBuilder.getId()
    }, { signal: new AbortController().signal });

    // Parse the response to get the ID to normalize
    const parsed = JSON.parse(result.content[0].text as string);
    const idToNormalize = parsed[0]?.id;

    // Normalize IDs and dates for snapshot testing
    const normalizedResult = createSnapshotResult(result, idToNormalize);

    // Verify the handler response using snapshot
    expect(normalizedResult).toMatchSnapshot();
  });

  it("should handle media type with no composition references", async () => {
    // Create a media type that isn't used as a composition
    const mediaTypeBuilder = await new MediaTypeBuilder()
      .withName(TEST_MEDIA_TYPE_NAME)
      .create();

    // Get the composition references
    const result = await GetMediaTypeCompositionReferencesTool().handler({
      id: mediaTypeBuilder.getId()
    }, { signal: new AbortController().signal });

    // Parse the response to get the ID to normalize
    const parsed = JSON.parse(result.content[0].text as string);
    const idToNormalize = parsed[0]?.id;

    // Normalize IDs and dates for snapshot testing
    const normalizedResult = createSnapshotResult(result, idToNormalize);

    // Verify the handler response using snapshot
    expect(normalizedResult).toMatchSnapshot();
  });
}); 