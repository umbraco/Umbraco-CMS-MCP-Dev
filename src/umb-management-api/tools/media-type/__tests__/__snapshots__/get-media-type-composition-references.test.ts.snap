// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-media-type-composition-references should get composition references for a media type 1`] = `
{
  "content": [
    {
      "text": "[{"id":"00000000-0000-0000-0000-000000000000","name":"_Test MediaType Composition","icon":"icon-folder"}]",
      "type": "text",
    },
  ],
}
`;

exports[`get-media-type-composition-references should handle media type with no composition references 1`] = `
{
  "content": [
    {
      "text": "[]",
      "type": "text",
    },
  ],
}
`;
